### ISSUE TITLE
cross compiling python modules for mitmproxy

### ISSUE THREAD

**Issue by cytec** [SynoCommunity/spksrc#2554]
@Dr-Bean @Diaoul i could need some support. I've recently started working on a `mitmproxy` SPK and therefore need to cross compile some python modules which require some native libraries... namely: `cryptography`, `Pillow`, `pyOpenSSL`, `ruamel.yaml` and `brotlipy` it seems like the first 3 already come with the Python SPK so i just skip those... 

however `brotlipy` seems to use `cffi` which simply won't compile at all... i've searched the web and found out that @Diaoul hat the same issue with `cryptography`

So what i've done now:
* created a new SPK and added `python` to `BUILD_DEPENDS`
* created a cross makefile for `ruamel.yaml` and `brotlipy` with `include ../../mk/spksrc.python-module.mk` based on `cross/cryptography`

I've got some errors when building the new cross packages, so i've installed both of them locally too. idk if that fixes it when moving the SPK to my DS. Next question is... if i require those cross modules do i still need to provide the wheels for those? if so... how? 

Any help would be appreciated since i have a feeling like i'm going nuts :D
My first guess was: hey ... why not just add those two (`ruamel.yaml` and `brotlipy`) to the python SPK in the first place... but again, also not sure if and how that would work...

**Comment by cytec** [comment]
Update: after reading https://github.com/SynoCommunity/spksrc/wiki/Using-wheels-to-distribute-Python-solutions again, i've changed the includes in `cross/ruamel.yaml` and `cross/brotlipy` to `include ../../mk/spksrc.python-wheel.mk` which then fails when compiling `cross/brotlipy` giving the error: `ImportError: /home/spksrc/spksrc/spk/mitmproxy/work-x86-5.2/install/usr/local/mitmproxy/lib/python2.7/site-packages/_cffi_backend.so: wrong ELF class: ELFCLASS64` which is what i remembered @Diaoul came across when building `cryptography`...

**Comment by Dr-Bean** [comment]
That doc is severely outdated, especially if you're on an updated fork: I've made a couple of changes.
It should be easy enough now. Take the Mercurial package as an example. Add those two projects to src/requirements.txt, and start compiling. The 'new' reason to create cross packages for Python stuff is either because the project needs patches, or because it depends on other cross-compiled packages (and other exceptions are when the project internally does funky stuff wrt dependency handling which it then shouldn't)

**Comment by cytec** [comment]
well the reason i made `cross` packages in the fist was that `ruamel.yaml` depends on `libyaml` and `brotlipy` depends on `cffi` which depends on `libffi` so i thought those should be compiled in order to create a working wheel?

**Comment by Dr-Bean** [comment]
Neither of the two complained when I created the wheels just now... 
But yeah, I suppose that's a reason to create a cross package instead. Let me see what happens here.

**Comment by Dr-Bean** [comment]
Got any code handy, by chance?

**Comment by cytec** [comment]
sure what do you need?

**Comment by Dr-Bean** [comment]
The Makefile contents would be nice ;)

**Comment by cytec** [comment]
spk/mitmproxy Makefile:

	SPK_NAME = mitmproxy
	SPK_VERS = $(shell date +%Y%m%d)
	SPK_REV = 1
	SPK_ICON = src/mitmproxy.png

	WHEELS = src/requirements.txt
	SPK_DEPENDS = "python>=2.7.11-15"

	BUILD_DEPENDS = cross/python cross/setuptools cross/pip cross/wheel
	DEPENDS = 

	MAINTAINER = cytec
	DESCRIPTION = mitmproxy is an interactive, SSL-capable intercepting proxy with a console interface.
	DISPLAY_NAME = mitmproxy
	BETA = 1
	CHANGELOG =

	HOMEPAGE   = https://mitmproxy.org/
	LICENSE    = MIT

	INSTALLER_SCRIPT = src/installer.sh
	SSS_SCRIPT       = src/dsm-control.sh

	INSTALL_PREFIX = /usr/local/$(SPK_NAME)

	POST_STRIP_TARGET = mitmproxy_extra_install

	# Pure Python package, make sure ARCH is not defined
	#override ARCH=

	include ../../mk/spksrc.spk.mk

	.PHONY: mitmproxy_extra_install
	mitmproxy_extra_install:
		install -m 755 -d $(STAGING_DIR)/var
		install -m 755 -d $(STAGING_DIR)/share
		install -m 755 -d $(STAGING_DIR)/share/wheelhouse
		install -m 644 $(WORK_DIR)/wheelhouse/* $(STAGING_DIR)/share/wheelhouse/


content of src/requirements.txt:

    blinker>=1.4, <1.5
    click>=6.2, <7
    certifi>=2015.11.20.1
    construct>=2.8, <2.9
    cssutils>=1.0.1, <1.1
    Flask>=0.10.1, <0.12
    h2>=2.5.0, <3
    html2text>=2016.1.8, <=2016.9.19
    hyperframe>=4.0.1, <5
    jsbeautifier>=1.6.3, <1.7
    passlib>=1.6.5, <1.8
    pyasn1>=0.1.9, <0.2
    pyparsing>=2.1.3, <2.2
    pyperclip>=1.5.22, <1.6
    requests>=2.9.1, <3
    tornado>=4.3, <4.5
    urwid>=1.3.1, <1.4
    watchdog>=0.8.3, <0.9
    sortedcontainers>=1.5.4, <1.6
    setuptools>=11.3, !=29.0.0
    mitmproxy==0.18.2
    brotlipy>=0.5.1, <0.7
    ruamel.yaml>=0.13.2, <0.14

**Comment by Dr-Bean** [comment]
It's very well possible that brotlipy doesn't have a build-time dependency on cffi btw, because cffi will be available via the Python package at runtime. That should mean my method works as intended.

`ruamel.yaml` is possibly a different case, because it directly relies on a normal cross package: that could very well be a build time dep.

**Comment by Dr-Bean** [comment]
ruamel.yaml's C stuff is optional...that's why it didn't complain. So you can go both ways wrt building the wheel for it. Either create a cross package like this:
```
PKG_NAME = ruamel.yaml
PKG_VERS = 0.13.4
PKG_EXT = tar.gz
PKG_DIST_NAME = $(PKG_NAME)-$(PKG_VERS).$(PKG_EXT)
PKG_DIST_SITE = https://files.pythonhosted.org/packages/source/r/$(PKG_NAME)
PKG_DIR = $(PKG_NAME)-$(PKG_VERS)

DEPENDS = cross/libyaml

HOMEPAGE =
COMMENT  =
LICENSE  =

include ../../mk/spksrc.python-wheel.mk
```
and add that to BUILD_DEPENDS of the SPK, and _remove_ it from src/requirements.

_or_

add it to src/requirements as you have there, and don't do anything else.

In both cases, make sure the PLIST for the spk contains `rsc:share/wheelhouse`, so the wheelhouse is included. See the last commit for the Mercurial package as example, it more or less shows the approach.

**Comment by cytec** [comment]
yeah! that's awesome... if that really works, i'd go the way without the cross package for now and check and test if the DS is capable of running mitmproxy in the first place... switching to the cross package shouldn't be that much of a deal if everything else is working properly.

Thanks for your help! i'll report back once the build process is done

**Comment by cytec** [comment]
Any idea why i get this: 

```  Could not find a version that satisfies the requirement netlib==0.17 (from -r /home/spksrc/syno/spk/mitmproxy/work-x64-5.2/wheelhouse/requirements.txt (line 21)) (from versions: 0.1, 0.2, 0.2.1, 0.2.2, 0.9, 0.9.1, 0.9.2, 0.10, 0.10.1, 0.11, 0.11.1, 0.11.2, 0.12.0, 0.12.1, 0.13, 0.13.1, 0.14.0, 0.15, 0.15.1, 0.16)
No matching distribution found for netlib==0.17 (from -r /home/spksrc/syno/spk/mitmproxy/work-x64-5.2/wheelhouse/requirements.txt (line 21))```

but there is a version 17 available... same goes for mitmproxy... it tells me latest version is 0.15 where it should be 0.18.2...

**Comment by Dr-Bean** [comment]
Netlib only has a wheel. We disallow downloading all wheels from PyPI, because pip isn't able to distinguish between the different types of wheels (pure-python, manylinux, what have you). Pip generally thinks that all wheels that are compatible with the host (so x64 in our case) are valid to use...which is obviously not the case, because we're cross-compiling for other arches.

Best course of action is request the maintainer to upload the source to PyPI next to the wheel (I do hope that just uoloading the wheel won't become common practice). For the time being, you can simply put the whole download link into requirements.txt.

**Comment by cytec** [comment]
was about to say something like that since i just checked the pypi page and saw that there is only a wheel file since those versions... 

nice to know one could provide a direct link to source! so i'll simply go with https://github.com/mitmproxy/mitmproxy/archive/v0.18.2.tar.gz till they upload source files to pypi again... if ever...

**Comment by Dr-Bean** [comment]
Maybe it's worth opening an issue on their tracker still, while informing them of an upcoming package ;) Besides that, I think PyPI also recommends uploading source files primarily, not just wheels, but of course I can't find the reference right now.

**Comment by cytec** [comment]
yeah i plan to do this once i've got it working ;)

**Comment by cytec** [comment]
@Dr-Bean is there any way of auto-depend-resolving or something like that? cause it seems like some packages are missing, installing them on DS works though so they are mostly pure python packages... but still: without installing them on the DS after installing the SPK you won't be able to run mitmproxy :/

**Comment by Dr-Bean** [comment]
Should prob add that to the wiki. We don't want auto dependency resolving on the targets. If the project doesn't supply a complete requirements-all, I generally just setup a virtualenv on the host (or you could do it on the DS as long as the deps are indeed pure-python), and then `pip install` the packages that are needed, without setting `--no-deps`. That means that pips dependency resolver will do its job and pull in any missing deps.
After that, run `pip freeze`, pipe it to a requirements.txt, and use that requirements.txt as a starting point to create the final src/requirements.txt.
Remove/comment out any and all deps that don't belong (setuptools, pip, wheel, python packages that are created via DEPENDS and python packages that come in via `--system-site-packages`, and you should end up with exactly what you need, pinned with correct versions and everything.

**Comment by cytec** [comment]
yup... thats what i did now... just thought that it would be cool if the host would resolve and download/build wheels for depends as well.

**Comment by Dr-Bean** [comment]
Nope, turned that off on purpose, at least right now. I _want_ you to make sure you've got the correct deps ;)

**Comment by cytec** [comment]
yeah... you are totally right... makes sense. where only a few packages so it wasn't that much to do ;)

hm well... thats strange... i've added `cross/pillow` to `BUILD_DEPENDS` and i get a wheel for it, but since i don't have it in requirements.txt like you said above... it seems like it gets totally ignored on install time... as a temp fix i've added `${INSTALL_DIR}/env/bin/pip install --no-deps --no-index -U --force-reinstall -f ${INSTALL_DIR}/share/wheelhouse pillow > /dev/null 2>&1
` to the installer... not sure if thats the right way or if anything is wrong on my end

**Comment by Dr-Bean** [comment]
No, the installer should install `*.whl`. Make sure you've got the new command, not the one that runs `-r requirements.txt`

**Comment by Dr-Bean** [comment]
Isn't pillow already in the Python package btw?

**Comment by cytec** [comment]
should be yes... but for some reason i have 3.0.0 and need 3.4.2 maybe my DS is fucked up a little or there is an older version included in the python spk :/

**Comment by Dr-Bean** [comment]
I didn't update the Python packages on the repo yet. You'll have to make sure to depend on the current version of the Python package, and build and install that before installing your package (applies to both Python 2.7 and 3.5)

**Comment by cytec** [comment]
yup i depend on `2.7.11-15` which is the last one on the repo but i've seen that spksrc has already a `2.7.12` more or less ready ;)

**Comment by Dr-Bean** [comment]
Yeah. Python 2.7.12 can be built and released, I just had other stuff going on. Can't update the Python3 package though, because that will break homeassistant...that package should prob be updated before releasing a new version. 
But with Python 3.6 and Python 2.7.13 both more or less around the corner wrt release dates, I was considering skipping the release, and update to latest in a couple of days.

**Comment by cytec** [comment]
seems like everything worked ;) installation went fine and i can start the package from command line...have not tested routing some traffic through it but i'll do when i get home.

btw since i compiled this still on the old vm (i686) `${INSTALL_DIR}/env/bin/pip install --no-deps --no-index -U --force-reinstall -f ${INSTALL_DIR}/share/wheelhouse ${INSTALL_DIR}/share/wheelhouse/*.whl` gave me an error since i have 2 versions of cross compiled wheels one named `cffi-1.9.1-cp27-cp27mu-linux_i686.whl` and one `cffi-1.9.1-cp27-none-any.whl` not quite sure why... as the cp statement https://github.com/SynoCommunity/spksrc/blob/master/mk/spksrc.wheel.mk#L72 here semms to copy `wheelhouse/cffi-1.9.1-cp27-cp27mu-linux_i686.whl` to `staging/share/wheelhouse/cffi-1.9.1-cp27-none-any.whl` where does the other one come from...

trying to install `cffi-1.9.1-cp27-cp27mu-linux_i686.whl` fails with `cffi-1.9.1-cp27-cp27mu-linux_i686.whl is not a supported wheel on this platform.` while isntalling `cffi-1.9.1-cp27-none-any.whl` runs perfectly fine... did i miss something? seems like those with an arch are duplicate/leftovers for some reason...

**Comment by Dr-Bean** [comment]
I haven't tested the wheel code on i686, but I guess it should work. Don't release packages with it though, some of the new arches can't handle it :)

Besides that, you seem to have added cffi to the requirements, and that shouldn't be necessary. It might also be that you added it to DEPENDS or created/edited the current cffi cross package?
Neither should be needed, because it's part of the Python package (well, if you've upgraded to 2.7.12, that is)

**Comment by Dr-Bean** [comment]
Maybe another possibility: do you have rsc:share/wheelhouse _and_ the copy of the wheelhouse dir in extra_install?

**Comment by Dr-Bean** [comment]
:) Don't mind that, using the SPKs PLIST stuff is new for the framework...it existed, but we never used it. But yeah, putting the wheelhouse stuff in extra_install is no longer needed with that. It opens up possibilities to add other dirs/files in the same way so we don't have to use extra_install as a catchall anymore.

**Comment by Dr-Bean** [comment]
As for cffi and pillow, if you prefer adding them as separate wheels, you can. The only thing to consider is when you use a virtualenv with `--system-site-packages`, pip might have trouble upgrading the existing modules to the later version, because it sees it as host packages and doesn't like touching them. That's when you can either choose to not set `--system-site-packages`, or hope for the best ;)

**Comment by cytec** [comment]
upgrading went fine... pip told me something like `upgrading xyz... not uninstalling xyz because it is outside of env` or something like that... so from my understanding it just installed the newer version for the env without touching the system one and went on with the next package... not 100% sure about that though...

well all of this isn't needed anymore once python > 2.7.11-15 is out ;) i'll keep it there for testing and remove it when the new python spk is released... if you get to that before i finish working on mitmproxy :)

**Comment by cytec** [comment]
@Dr-Bean well... seems like they don't plan on uploading the source anymore. as stated in the issue i linked above: 

```
We do not plan to publish source distributions on pypi, as those would be
picked up by pip2. Our next release will be Python 3 only and we prefer to
have a clear "no distribution found" message for Python 2 users then. Sorry.
```

so i guess we have to stick with either using the source link to github in requirements.txt or maybe create a separate cross package?

**Comment by cytec** [comment]
Besides that: since i'd like to provide this as a CLI only SPK (also there is a webinterface available after installation) which installs 3 binaries to `/usr/local/mitmproxy/env/bin/` i was wondering if we should symlink them to `/usr/local/mitmproxy/bin` or something like that. Or state the install path in the Description?

installed `bin` files are:
* mitmproxy
* mitmweb
* mitmdump

Or do we go the other way and use `mitmweb` as default and make it startable via Package Center and a custom config file? if se we would need at least 2 Ports (one for the proxy and one for the web interface) and i have to digg in if we can change the default `~/.mitmproxy` directory somehow... i know that works for the certificates used by adding `--cadir` i'd guess that there is an option to provide a config file somehow too... what are you'r thoughts?

**Comment by Dr-Bean** [comment]
I'll take a look at the PR later. Unfortunate that the maintainers won't upload a tarball, but they have a point: I don't think pip has the ability to check that requirement (unless you count failing at first run, and catching that...). I believe something is in development for Warehouse (PyPI's new codebase), but if it has to trickle down to pip and setuptools, it could be a while.

If you use the link in requirements.txt, you can rewrite it in extra_install. See the ffsync package, I did the same thing there. The other option would be the cross package (e.g. cross/yenc)

Having said that, if mitmproxy is moving to Python3, then maybe it makes sense to depend on the Python3 package, instead of on the Py2 package? Shouldn't be too hard to change, I don't think.

For the `~/.mitmproxy` location, note that DSM6's users have an unchangeable homedir, which is set to `/var/packages/${PACKAGE}/target` (so, no `/var` subdir), so that may be something to look out for.
As for startable vs CLI only, that's your choice. I have no problem with CLI, especially if users ultimately still have a choice to use a web interface if they so choose.

How do we handle symlinks to other packages with their binaries in a venv? Don't we link directly to `/usr/local/bin` from the virtualenv? I don't mind that...I think Mercurial at the very least does that. Maybe you'd just want to link the main binary though, the others are for advanced usage imo.

**Comment by Dr-Bean** [comment]
@cytec Maybe you've already seen this: https://corte.si/posts/code/mitmproxy/announce_1_0/index.html? 
Should be Python3 with that if I'm not mistaken. On the use of the web interface, I came across this a couple of days ago: https://news.ycombinator.com/item?id=13259329, so preferring CLI makes sense from that pov.

**Comment by Dr-Bean** [comment]
@cytec 
Another solution for `netlib` only having a wheel on PyPI and no sdist: we can allow using that wheel specifically in the requirements file by doing this:
```
--only-binary=netlib # Source not on PyPI
netlib==0.17
```
What this does is override `PIP_WHEEL`'s setting of `--no-binary :all:` _just_ for netlib. Note that this can only be used for universal wheels (wheels with `-py2.py3-none-any.whl` naming format, although it works for `py2-none-any` and `py3-none-any` too)

Ran into a similar issue for three other modules. Seems it's a better solution than using a source download...or at least easier ;)

**Comment by hgy59** [comment]
We never created a mitmproxy package, but with the current python3 support it should be possible.
Please create a new issue if mitmproxy is still requested.
